File: ruby/GrabzItDemo/app/controllers/home_controller.rb
require 'URI'
require 'grabzit'

class HomeController < ActionController::Base
  def index
	@UseCallbackHandler = useCallbackHandler(request)
  end
  
  def useCallbackHandler(r)
	app_config = YAML.load_file("#{Rails.root}/config/config.yml")[Rails.env]
	if app_config['handler'] !~ URI::regexp
		return false
	end
	localhost = Regexp.union [/^127\.\d{1,3}\.\d{1,3}\.\d{1,3}$/, /^::1$/, /^0:0:0:0:0:0:0:1(%.*)?$/]
	return localhost !~ r.remote_addr && localhost !~ r.remote_ip	
  end
  
  def processScreenshot
  	url = params[:url]
	html = params[:html]
  	format = params[:format]
	isHtml = params[:convert] == "html"
  	
  	app_config = YAML.load_file("#{Rails.root}/config/config.yml")[Rails.env]
  	grabzItClient = GrabzIt::Client.new(app_config['application_key'], app_config['application_secret'])
  	
  	if format == "jpg"
		if isHtml
			grabzItClient.html_to_image(html)
		else
			grabzItClient.url_to_image(url)
		end
	elsif format == "docx"
		if isHtml
			grabzItClient.html_to_docx(html)
		else
			grabzItClient.url_to_docx(url)
		end		
	elsif format == "csv"
		if isHtml
			grabzItClient.html_to_table(html)
		else
			grabzItClient.url_to_table(url)
		end
  	elsif format == "gif"
	  	grabzItClient.url_to_animation(url)
  	else
		if isHtml
			grabzItClient.html_to_pdf(html)
		else
			grabzItClient.url_to_pdf(url)
		end
  	end
  	
  	@SuccessMessage = ''
  	@ErrorMessage = ''
  	
  	begin
		if useCallbackHandler(request)
			grabzItClient.save(app_config['handler'])
			@SuccessMessage = "Processing..."
			@UseCallbackHandler = true
		else
			grabzItClient.save_to("public/screenshots/"+rand(1...999999).to_s()+"."+format)
			@UseCallbackHandler = false
		end
	rescue RuntimeError, GrabzIt::GrabzItException => e
		@ErrorMessage = e
	end
	
	render :action => "index.html"
  end
  
  def clearScreenshots
  	dir_path = "public/screenshots/"
  	Dir.foreach(dir_path) {|f| fn = File.join(dir_path, f); File.delete(fn) if f != '.' && f != '..'}
  	
  	redirect_to root_path
  end
end

File: aspnet/SampleMVC/Controllers/HomeController.cs
using GrabzIt;
using GrabzIt.Screenshots;
using SampleMVC.Models;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.IO;
using System.Web.Mvc;

namespace SampleMVC.Controllers
{
    public class HomeController : Controller
    {
        [HttpPost]
        [ValidateInput(false)]
        public ActionResult Index(FormModel form)
        {
            GrabzItClient grabzItClient = GrabzItClient.Create(ConfigurationManager.AppSettings["ApplicationKey"], ConfigurationManager.AppSettings["ApplicationSecret"]);

            form.UseCallbackHandler = !HttpContext.Request.IsLocal;
            try
            {
                if (form.Format == "jpg")
                {
                    if (form.Convert == "html")
                    {
                        grabzItClient.HTMLToImage(form.HTML);
                    }
                    else
                    {
                        grabzItClient.URLToImage(form.URL);
                    }
                }
                else if (form.Format == "docx")
                {
                    if (form.Convert == "html")
                    {
                        grabzItClient.HTMLToDOCX(form.HTML);
                    }
                    else
                    {
                        grabzItClient.URLToDOCX(form.URL);
                    }
                }
                else if (form.Format == "csv")
                {
                    if (form.Convert == "html")
                    {
                        grabzItClient.HTMLToTable(form.HTML);
                    }
                    else
                    {
                        grabzItClient.URLToTable(form.URL);
                    }
                }
                else if (form.Format == "gif")
                {
                    grabzItClient.URLToAnimation(form.URL);
                }
                else
                {
                    if (form.Convert == "html")
                    {
                        grabzItClient.HTMLToPDF(form.HTML);
                    }
                    else
                    {
                        grabzItClient.URLToPDF(form.URL);
                    }
                }
                if (form.UseCallbackHandler)
                {
                    grabzItClient.Save(HttpContext.Request.Url.Scheme + "://" + HttpContext.Request.Url.Authority + HttpContext.Request.ApplicationPath + "Home/Handler");                    
                }
                else
                {
                    grabzItClient.SaveTo(Server.MapPath("~/results/" + Guid.NewGuid().ToString() + "." + form.Format));
                }
                form.Message = "Processing...";
            }
            catch (Exception ex)
            {
                form.Error = ex.Message;
            }
            return View(form);
        }

        public ActionResult Handler(string filename, string id, string message, string customId, string format, int targeterror)
        {
            GrabzItClient grabzItClient = GrabzItClient.Create(ConfigurationManager.AppSettings["ApplicationKey"], ConfigurationManager.AppSettings["ApplicationSecret"]);
            GrabzItFile file = grabzItClient.GetResult(id);
            //Ensure that the application has the correct rights for this directory.
            file.Save(Server.MapPath("~/results/" + filename));

            return null;
        }

        public ActionResult Index()
        {
            FormModel form = new FormModel();
            form.UseCallbackHandler = !HttpContext.Request.IsLocal;
            return View(form);
        }

        public JsonResult Results()
        {
            string[] files = Directory.GetFiles(Server.MapPath("~/results"));
            List<string> results = new List<string>();
            foreach (string file in files)
            {
                if (string.IsNullOrEmpty(file) || file.Contains(".txt"))
                {
                    continue;
                }
                FileInfo info = new FileInfo(file);
                results.Add("../results/" + info.Name);
            }

            return this.Json(results, JsonRequestBehavior.AllowGet);
        }

        public ActionResult Delete()
        {
            string[] files = Directory.GetFiles(Server.MapPath("~/results"));
            foreach (string file in files)
            {
                System.IO.File.Delete(file);
            }

            return RedirectToAction("Index");
        }
    }
}